### PR TITLE
feat: new data structure saved in keychain

### PR DESCRIPTION
**Pull request by sonle-geekyants**
**- What I did**
Upgrade data structure saved in Keychain to fix this ticket [#468](https://github.com/atsign-foundation/apps/issues/468)
**- How I did it**

**- How to verify it**

**- Description for the changelog**

### CODE REVIEW

File: at_client_mobile/example/lib/src/screens/login.dart [hunk 1]
```diff
@@ -50,33 +52,89 @@ class _LoginScreen extends State<LoginScreen> {
...
+                  //     clientSDKInstance.atClientServiceInstance = value[atsign];
+                  //     _logger.finer('Successfully onboarded $atsign');
+                  //   },
+                  //   onError: (Object? error) {
+                  //     // _logger.severe('Onboarding throws $error error');
+                  //   },
+                  //   nextScreen: const HomeScreen(),
+                  //   appAPIKey: AtEnv.appApiKey,
+                  //   rootEnvironment: AtEnv.rootEnvironment,
+                  // );
```
Comment (gkc) [inline at_client_mobile/example/lib/src/screens/login.dart]: Please delete dead code.

File: at_client_mobile/lib/src/at_client_auth.dart [hunk 2]
```diff
@@ -41,12 +42,13 @@ class AtClientAuthenticator implements AtClientAuth {
...
-        await _keyChainManager.getValue(atSign, keychainPKAMPrivateKey);
-    var encryptionPrivateKey =
-        await _keyChainManager.getValue(atSign, keychainEncryptionPrivateKey);
+    final atsign = await _keyChainManager.readAtsign(name: atSign);
+    var publicKey = atsign?.pkamPublicKey;
+        // await _keyChainManager.getValue(atSign, keychainPKAMPublicKey);
+    var privateKey = atClientPreference.privateKey ??= atsign?.pkamPrivateKey;
+        // await _keyChainManager.getValue(atSign, keychainPKAMPrivateKey);
+    var encryptionPrivateKey =  atsign?.encryptionPublicKey;
+        // await _keyChainManager.getValue(atSign, keychainEncryptionPrivateKey);
```
Comment (gkc) [inline at_client_mobile/lib/src/at_client_auth.dart]: please delete dead code

File: at_client_mobile/lib/src/at_client_auth.dart [hunk 3]
```diff
@@ -96,14 +98,21 @@ class AtClientAuthenticator implements AtClientAuth {
...
-          await _keyChainManager.putValue(
-              atSign, keychainSelfEncryptionKey, selfEncryptionKey);
+          // await _keyChainManager.putValue(
+          //     atSign, keychainSelfEncryptionKey, selfEncryptionKey);
+          atSignItem = atSignItem.copyWith(
+            encryptionPrivateKey: encryptionKeyPair.privateKey.toString(),
+            encryptionPublicKey: encryptionPubKey,
+            selfEncryptionKey: selfEncryptionKey,
+          );
+          await _keyChainManager.storeAtSign(atSign: atSignItem);
```
Comment (gkc) [inline at_client_mobile/lib/src/at_client_auth.dart:104]: again, please delete dead code above

File: at_client_mobile/lib/src/at_client_service.dart [hunk 4]
```diff
@@ -195,21 +196,39 @@ class AtClientService {
     var encryptionPublicKey = EncryptionUtil.decryptValue(
         extractedjsonData[BackupKeyConstants.ENCRYPTION_PUBLIC_KEY_FROM_FILE],
         decryptKey);
-    await _keyChainManager.putValue(
-        atsign, keychainEncryptionPublicKey, encryptionPublicKey);
+    // await _keyChainManager.putValue(
+    //     atsign, keychainEncryptionPublicKey, encryptionPublicKey);
```
Comment (gkc) [inline at_client_mobile/lib/src/at_client_service.dart]: delete dead code

File: at_client_mobile/lib/src/at_client_service.dart [hunk 5]
```diff
@@ -195,21 +196,39 @@ class AtClientService {
...
         decryptKey);
-    await _keyChainManager.putValue(
-        atsign, keychainEncryptionPrivateKey, encryptionPrivateKey);
-    await _keyChainManager.putValue(
-        atsign, keychainSelfEncryptionKey, decryptKey);
+    // await _keyChainManager.putValue(
+    //     atsign, keychainEncryptionPrivateKey, encryptionPrivateKey);
+    // await _keyChainManager.putValue(
+    //     atsign, keychainSelfEncryptionKey, decryptKey);
+
```
Comment (gkc) [inline at_client_mobile/lib/src/at_client_service.dart:203]: delete dead code

File: at_client_mobile/lib/src/at_client_service.dart [hunk 6]
```diff
@@ -250,8 +269,10 @@ class AtClientService {
   ///Returns [OnboardingStatus] of the atsign by checking it with remote server.
   Future<OnboardingStatus> getKeyRestorePolicy(String atSign) async {
     var serverEncryptionPublicKey = await _getServerEncryptionPublicKey(atSign);
+    // var localEncryptionPublicKey =
+    //     await _keyChainManager.getValue(atSign, keychainEncryptionPublicKey);
```
Comment (gkc) [inline at_client_mobile/lib/src/at_client_service.dart]: delete dead code

File: at_client_mobile/lib/src/at_client_service.dart [hunk 7]
```diff
@@ -359,14 +380,16 @@ class KeychainUtil {
   }
 
   static Future<String?> getAESKey(String atsign) async {
-    return await _keyChainManager.getValue(atsign, keychainSelfEncryptionKey);
+    // return await _keyChainManager.getValue(atsign, keychainSelfEncryptionKey);
```
Comment (gkc) [inline at_client_mobile/lib/src/at_client_service.dart]: delete dead code

**Review by gkc — CHANGES_REQUESTED** [review]
Looks OK to me but I'm not expert in the flutter world. Perhaps @XavierChanth you can take a look at this PR also?

My main nit with the PR is that there is a lot of dead code which has been commented-out rather than deleted from the files. Please delete dead code. We use our version control systems to see code diffs. When dead code is removed, then the diffs are easier to read for reviewers, and we don't need a second PR later simply to remove the commented-out code.

**Comment by yahu1031** [comment]
I have used this branch and tested it with the new flutter project. But looks like we still need to fix flutter_keychain first. But we can still push this code to trunk but not good for publish.

File: at_client_mobile/lib/src/at_client_auth.dart [hunk 8]
```diff
@@ -1,10 +1,11 @@
 import 'package:at_client_mobile/at_client_mobile.dart';
-import 'package:at_client_mobile/src/auth_constants.dart';
 import 'package:at_commons/at_builders.dart';
 import 'package:at_lookup/at_lookup.dart';
 import 'package:at_utils/at_logger.dart';
 import 'package:crypton/crypton.dart';
 
+import 'atsign_key.dart';
```
Comment (xavierchanth) [inline at_client_mobile/lib/src/at_client_auth.dart]: Try to be consistent with import conventions.

File: at_client_mobile/lib/src/at_client_service.dart [hunk 9]
```diff
@@ -2,11 +2,12 @@ import 'dart:convert';
 import 'dart:core';
 
 import 'package:at_client_mobile/at_client_mobile.dart';
-import 'package:at_client_mobile/src/auth_constants.dart';
 import 'package:at_commons/at_builders.dart';
 import 'package:at_lookup/at_lookup.dart';
 import 'package:at_utils/at_logger.dart';
 
+import 'atsign_key.dart';
```
Comment (xavierchanth) [inline at_client_mobile/lib/src/at_client_service.dart]: import conventions

File: at_client_mobile/lib/src/keychain_manager.dart [hunk 10]
```diff
@@ -24,27 +26,345 @@ class KeyChainManager {
...
+  /// Function to get atsign's key with name
+  Future<AtsignKey?> readAtsign({required String name}) async {
+    final atSigns = await readAtsigns();
+    if (atSigns.isNotEmpty) {
+      for (int i = 0; i < atSigns.length; i++) {
+        if (atSigns[i].name == name) {
+          return atSigns[i];
+        }
+      }
+      return null;
```
Comment (xavierchanth) [inline at_client_mobile/lib/src/keychain_manager.dart:256]: delete redundant return statement

File: at_client_mobile/lib/src/atsign_key.dart [hunk 11]
```diff
@@ -0,0 +1,152 @@
+///Save atsign key
+/// https://docs.google.com/document/d/1JAXNrGr6J30m1xTWD4t7z2eQRo6O7icEOprJh1KKNas/edit?hl=en&forcehl=1#
+class AtsignKey {
+  final String name;
```
Comment (xavierchanth) [inline at_client_mobile/lib/src/atsign_key.dart]: 'name' should really be called 'atSign'

**Review by xavierchanth — CHANGES_REQUESTED** [review]

**Comment by xavierchanth** [comment]
Please try to use more guard clauses, this will increase readability of the code.

**Comment by sonle-geekyants** [comment]
Thanks, @gkc @yahu1031, and @XavierChanth 
I've pushed a new commit to solve your comments. please help me check again!

File: at_client_mobile/example/lib/src/screens/login.dart [hunk 12]
```diff
@@ -50,33 +51,69 @@ class _LoginScreen extends State<LoginScreen> {
...
                   );
+                  switch (result) {
+                    case AtOnboardingResult.success:
+                      Navigator.push(
+                          context,
+                          MaterialPageRoute(
+                              builder: (_) => const HomeScreen()));
+                      break;
+                    case AtOnboardingResult.error:
+                    // TODO: Handle this case.
```
Comment (gkc) [inline at_client_mobile/example/lib/src/screens/login.dart]: You need to ... handle this case!

File: at_client_mobile/example/lib/src/screens/login.dart [hunk 13]
```diff
@@ -50,33 +51,69 @@ class _LoginScreen extends State<LoginScreen> {
...
+                      break;
+                    case AtOnboardingResult.cancel:
+                    // TODO: Handle this case.
```
Comment (gkc) [inline at_client_mobile/example/lib/src/screens/login.dart]: You need to ... handle this case!

File: at_client_mobile/example/lib/src/screens/login.dart [hunk 14]
```diff
@@ -50,33 +51,69 @@ class _LoginScreen extends State<LoginScreen> {
...
+                  );
+                  switch (result) {
+                    case AtOnboardingResult.success:
+                      Navigator.push(
+                          context,
+                          MaterialPageRoute(
+                              builder: (_) => const HomeScreen()));
+                      break;
+                    case AtOnboardingResult.error:
+                      // TODO: Handle this case.
```
Comment (gkc) [inline at_client_mobile/example/lib/src/screens/login.dart]: You need to ... handle this case!

File: at_client_mobile/example/lib/src/screens/login.dart [hunk 15]
```diff
@@ -50,33 +51,69 @@ class _LoginScreen extends State<LoginScreen> {
...
+                      break;
+                    case AtOnboardingResult.cancel:
+                      // TODO: Handle this case.
```
Comment (gkc) [inline at_client_mobile/example/lib/src/screens/login.dart]: You need to ... handle this case!

File: at_client_mobile/lib/src/at_client_auth.dart [hunk 16]
```diff
@@ -96,14 +94,15 @@ class AtClientAuthenticator implements AtClientAuth {
...
+          var atSignItem = await _keyChainManager.readAtsign(name: atSign) ?? AtsignKey(atSign: atSign);
           var encryptionPubKey = encryptionKeyPair.publicKey.toString();
-          await _keyChainManager.putValue(
-              atSign, keychainEncryptionPublicKey, encryptionPubKey);
           var selfEncryptionKey = EncryptionUtil.generateAESKey();
-          await _keyChainManager.putValue(
-              atSign, keychainSelfEncryptionKey, selfEncryptionKey);
+          atSignItem = atSignItem.copyWith(
+            encryptionPrivateKey: encryptionKeyPair.privateKey.toString(),
+            encryptionPublicKey: encryptionPubKey,
```
Comment (gkc) [inline at_client_mobile/lib/src/at_client_auth.dart]: For consistency, please replace the above line with `encryptionPublicKey: encryptionKeyPair.publicKey.toString();` (and remove the now unused `encryptionPubKey` local variable)

**Review by gkc — CHANGES_REQUESTED** [review]
Thanks Son for deleting all that dead code. Much easier to follow the diffs now.

I've added some requests for a few new changes

**Comment by sonle-geekyants** [comment]
Thanks @gkc 
I've upgraded the example project with the newest flutter version and corrected the code base on your comments. Please help me check it again!

**Comment by sonle-geekyants** [comment]
Hi @gkc and @XavierChanth,
Could you please recheck the PR?

**Review by xavierchanth — APPROVED** [review]

**Comment by sonle-geekyants** [comment]
@gkc please approve this PR and we will prepare to release it.

**Review by gkc — APPROVED** [review]
👍👍

**Pull request merged**